app/battle.py

```diff
@@ -0,0 +1,34 @@
+from typing import Dict, Any
+from app.knight import Knight
+
+
+def battle(knights_config: Dict[str, Any]) -> Dict[str, int]:
+    lancelot = Knight(knights_config["lancelot"])
+    arthur = Knight(knights_config["arthur"])
+    mordred = Knight(knights_config["mordred"])
+    red_knight = Knight(knights_config["red_knight"])
+
+    # Battle: Lancelot vs Mordred
+    lancelot.hp -= max(0, mordred.power - lancelot.armour.protection)
+    mordred.hp -= max(0, lancelot.power - mordred.armour.protection)
+
+    if lancelot.hp <= 0:
+        lancelot.hp = 0
+    if mordred.hp <= 0:
+        mordred.hp = 0
+
+    # Battle: Arthur vs Red Knight
+    arthur.hp -= max(0, red_knight.power - arthur.armour.protection)
+    red_knight.hp -= max(0, arthur.power - red_knight.armour.protection)
+
+    if arthur.hp <= 0:
+        arthur.hp = 0
+    if red_knight.hp <= 0:
+        red_knight.hp = 0
```

Comment:
The code for the battle between two knights is repeated twice. This violates the DRY principle. Instead of repeating the code, you could create a separate function that handles the fight between two knights. It would take two knights as arguments and update their hp based on their power and armour protection. This way, you can call this function for each pair of knights, reducing code duplication.